### PR TITLE
Introduce a tier-1 JIT compiler based on x86-64 architecture

### PR DESCRIPTION
**Pull request by qwe661234**
When the using frequency of a block exceeds a predetermined threshold, the tier-1 JIT compiler traces the chained block and generate  corresponding low quailty machine code. The resulting target machine code is stored in the code cache for future utilization.

The primary objective of introducing the tier-1 JIT compiler is to enhance the execution speed of RISC-V instructions. This implementation requires two additional components: a tier-1 machine code generator,  and code cache. Furthermore, this tier-1 JIT compiler serves as the foundational target for future improvements.

In addition, we have developed a Python script that effectively traces code templates and automatically generates JIT code templates. This approach eliminates the need for manually writing duplicated code.

As shown in the performance analysis below, the tier-1 JIT compiler's  performance closely parallels that of QEMU in benchmarks with a  constrained dynamic instruction count. However, for benchmarks  featuring a substantial dynamic instruction count or lacking specific hotspots—examples include pi and STRINGSORT—the tier-1 JIT compiler demonstrates noticeably slower execution compared to QEMU.

Hence, a robust tier-2 compiler is essential to generate optimized machine code across diverse execution paths, coupled with a runtime profiler for detecting hotspots.

* Performance

| Metric   | rv32emu(T1C) | qemu  |
|----------|-----------------|-------|
|aes	   |             0.02|  0.031|
|mandelbrot|	        0.029| 0.0115|
|puzzle	   |           0.0115|	0.009|
|pi	       |           0.0413| 0.0177|
|dhrystone |	        0.331|	0.393|
|Nqeueens  |	        0.854|	0.749|
|qsort-O2  |	        2.384|	 2.16|
|miniz-O2  |	         1.33|	 1.01|
|primes-O2 |	         2.93|	1.069|
|sha512-O2 |	        2.057|	0.939|
|stream	   |           12.747|	10.36|
|STRINGSORT|	       89.012| 11.496|

As demonstrated in the memory usage analysis below, the tier-1 JIT compiler utilizes less memory than QEMU across all benchmarks.

* Memory usage

| Metric   | rv32emu(T1C) |   qemu  |
|----------|-----------------|---------|
|aes	   |          186,228|1,343,012|
|mandelbrot|	      152,203|  841,841|
|puzzle	   |          153,423|  890,225|
|pi	       |          152,923|  879,957|
|dhrystone |	      154,466|  856,404|
|Nqeueens  |	      154,880|  858,618|
|qsort-O2  |	      155,091|  933,506|
|miniz-O2  |	      165,627|1,076,682|
|primes-O2 |	      150,540|  928,446|
|sha512-O2 |	      153,553|	978,177|
|stream	   |          165,911|  957,845|
|STRINGSORT|	      167,871|1,104,702|

Related: #238

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
Eliminate the ubpf specific code. In fact, you already construct a new JIT backend.

File: src/riscv_private.h [hunk 1]
```diff
@@ -116,11 +110,32 @@ struct riscv_internal {
     uint32_t csr_mip;      /* Machine interrupt pending */
     uint32_t csr_mbadaddr;
 
-    bool compressed;       /**< current instruction is compressed or not */
+    bool compressed; /**< current instruction is compressed or not */
+#if !RV32_HAS(JIT)
     block_map_t block_map; /**< basic block map */
+#else
+    struct cache *block_cache;
+#ifdef MIR
```
Comment (jserv) [inline src/riscv_private.h]: Drop MIR specific code.

File: src/rv32_template.c [hunk 2]
```diff
@@ -4,42 +4,73 @@
...
-RVOP(nop, { rv->X[rv_reg_zero] = 0; })
+RVOP(nop, { rv->X[rv_reg_zero] = 0; }, {RVCODEGEN(nop, {})})
 
 /* LUI is used to build 32-bit constants and uses the U-type format. LUI
  * places the U-immediate value in the top 20 bits of the destination
  * register rd, filling in the lowest 12 bits with zeros. The 32-bit
  * result is sign-extended to 64 bits.
  */
-RVOP(lui, { rv->X[ir->rd] = ir->imm; })
+RVOP(lui, { rv->X[ir->rd] = ir->imm; }, {RVCODEGEN(lui, {
```
Comment (jserv) [inline src/rv32_template.c]: Instead of `RVCODEGEN`, use `X64` identifier instead. Later, we may integrate A64 backend as well.

File: src/rv32_template.c [hunk 3]
```diff
@@ -4,42 +4,73 @@
...
+         emit_load_imm(state, RAX, ir->imm);
+         emit_store(state, S32, RAX, platform_parameter_registers[0],
```
Comment (jserv) [inline src/rv32_template.c]: Replace `platform_parameter_registers[0]` with something shorter.

File: src/rv32_template.c [hunk 4]
```diff
@@ -4,42 +4,73 @@
...
+                    offsetof(struct riscv_internal, X) + 4 * ir->rd);
```
Comment (jserv) [inline src/rv32_template.c]: I anticipate the presence of template-based code generation, which necessitates minimal descriptions of target instructions without the need for complete C source code.

File: src/rv32_template.c (hunk 2 above)
Comment (jserv) [reply src/rv32_template.c]: Can you avoid the repeated `lui` declaration?

File: src/cache.c [hunk 5]
```diff
@@ -3,20 +3,19 @@
...
-
-#include "cache.h"
 #include "mpool.h"
 #include "utils.h"
 
 /* THRESHOLD is set to identify hot spots. Once the frequency of use for a block
  * exceeds the THRESHOLD, the JIT compiler flow is triggered.
  */
-#define THRESHOLD 1000
+#define THRESHOLD 4096
```
Comment (jserv) [inline src/cache.c:22]: Add `FIXME` to indicate the need of an effective (dynamic) profiler.

File: src/rv32_template.c (hunk 2 above)
Comment (qwe661234) [reply src/rv32_template.c]: > Can you avoid the repeated `lui` declaration?

Does you mean rewrite to `RVCODEGEN({})`?

File: src/rv32_template.c (hunk 4 above)
Comment (qwe661234) [reply src/rv32_template.c]: Does you mean writing IR here, and this IR would be parsed and translatedin to the current C code by python script?

File: src/rv32_template.c (hunk 2 above)
Comment (qwe661234) [reply src/rv32_template.c:14]: > Instead of `RVCODEGEN`, use `X64` identifier instead. Later, we may integrate A64 backend as well.

Actually, every lines of machine code are generated from API, so we can switch the target machine in the API.

Comment (jserv) [reply src/rv32_template.c]: I expect the instruction such as "lui" appears once within the scope of `RVOP` instead of twice.

File: src/rv32_template.c (hunk 4 above)
Comment (jserv) [reply src/rv32_template.c]: Exactly. The template-based JIT code generation is mapped to the use of a custom domain-specific language (DSL), which is interpreted by our Python script. Having minimal descriptions of target instructions without the need for complete C source code makes sense for both maintainability and future machine code lifting. See [Rellume — Lift machine code to LLVM IR](https://github.com/aengelke/rellume),

Comment (jserv) [reply src/rv32_template.c]: Meanwhile, you can include a brief documentation section at the beginning of `rv32_template.c` explaining the use of DSL during JIT code generation.

File: Makefile [hunk 6]
```diff
@@ -118,6 +118,22 @@ gdbstub-test: $(BIN)
 	$(Q).ci/gdbstub-test.sh && $(call notice, [OK])
 endif
 
+ENABLE_JIT ?= 0
+$(call set-feature, JIT)
+ifeq ($(call has, JIT), 1)
+OBJS_EXT += ubpf_jit_x86_64.o 
+uname_p := $(shell uname -p)
+ifneq ($(uname_p), x86_64)
+$(error JIT mode only support for x86_64 target now.)
```
Comment (jserv) [inline Makefile]: Improve the grammar.

**Review by jserv — CHANGES_REQUESTED** [review]
After removing the outdated ubpf-jit code, rename `ubpf_jit_x86_64.[ch]` to `jit-x64.[ch]`, which serves as the infrastructure for the tier-1 JIT compiler.

File: src/jit_x64.c [hunk 7]
```diff
@@ -0,0 +1,745 @@
...
+    BPF_REG_2,
+    BPF_REG_3,
+    BPF_REG_4,
+    BPF_REG_5,
+    BPF_REG_6,
+    BPF_REG_7,
+    BPF_REG_8,
+    BPF_REG_9,
+    BPF_REG_10,
+    _BPF_REG_MAX,
```
Comment (jserv) [inline src/jit_x64.c]: Refine these names for identifiers.

File: src/jit_x64.c [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+#define EBPF_SIZE_W 0x00
+#define EBPF_SIZE_H 0x08
+#define EBPF_SIZE_B 0x10
+#define EBPF_SIZE_DW 0x18
+
+/* Other memory modes are not yet supported */
+#define EBPF_MODE_IMM 0x00
+#define EBPF_MODE_MEM 0x60
+
+#define EBPF_OP_ADD_IMM (EBPF_CLS_ALU | EBPF_SRC_IMM | 0x00)
```
Comment (jserv) [inline src/jit_x64.c:79]: Ditto. Use consistent naming.

File: src/rv32_template.c [hunk 9]
```diff
@@ -3,43 +3,75 @@
...
  */
 
+/* Please check files gen-jit-template.py and jit_x64.[ch] to understand custom
+ * domain-specific language of tiered 1 JIT compiler, ecah opcode has
+ * corresonding API. For example, the API of `ld_imm` is `emit_load_imm`.
+ */
+
 /* Internal */
-RVOP(nop, { rv->X[rv_reg_zero] = 0; })
+RVOP(nop, { rv->X[rv_reg_zero] = 0; }, {X64({})})
```
Comment (jserv) [inline src/rv32_template.c]: How about using shorter form? That is,
```c
RVOP(nop, { rv->X[rv_reg_zero] = 0; }, X64 { } )
```

For interpreter path, we can define a dummy `X64` macro which ignores everything.

Comment (qwe661234) [reply src/rv32_template.c:12]: > How about using shorter form? That is,
> 
> ```c
> RVOP(nop, { rv->X[rv_reg_zero] = 0; }, X64 { } )
> ```
> 
> For interpreter path, we can define a dummy `X64` macro which ignores everything.

It can only be (X64{}), the compiler would print error for X64 {} because the MACRO of RVOP would extend incorrectly.

Comment (qwe661234) [reply src/rv32_template.c]: The compiler would print error for X64 {} because the MACRO of RVOP are extended incorrectly.

Comment (jserv) [reply src/rv32_template.c]: We can still stick to the shorter form `RVOP(nop, { rv->X[rv_reg_zero] = 0; }, X64({ }) )`, right?

File: Makefile [hunk 10]
```diff
@@ -118,6 +118,22 @@ gdbstub-test: $(BIN)
 	$(Q).ci/gdbstub-test.sh && $(call notice, [OK])
 endif
 
+ENABLE_JIT ?= 0
+$(call set-feature, JIT)
+ifeq ($(call has, JIT), 1)
+OBJS_EXT += jit_x64.o 
+uname_p := $(shell uname -p)
+ifneq ($(uname_p), x86_64)
```
Comment (jserv) [inline Makefile]: Reuse the variable `processor` defined in `mk/toolchain.mk`.

File: Makefile [hunk 11]
```diff
@@ -118,6 +118,22 @@ gdbstub-test: $(BIN)
 	$(Q).ci/gdbstub-test.sh && $(call notice, [OK])
 endif
 
+ENABLE_JIT ?= 0
+$(call set-feature, JIT)
+ifeq ($(call has, JIT), 1)
+OBJS_EXT += jit_x64.o 
+uname_p := $(shell uname -p)
```
Comment (jserv) [inline Makefile]: Drop unnecessary `uname -p`.

File: src/rv32_template.c (hunk 9 above)
Comment (qwe661234) [reply src/rv32_template.c]: This form is OK.

**Review by jserv — CHANGES_REQUESTED** [review]
Reuse processor detection in build system.

File: src/common.h [hunk 12]
```diff
@@ -6,7 +6,6 @@
 #pragma once
 
 #include "feature.h"
-
```
Comment (jserv) [inline src/common.h]: Don't change this line.

File: src/jit_x64.c (hunk 7 above)
Comment (jserv) [reply src/jit_x64.c]: Replace all identifiers which contain BPF with the refined ones.

File: src/jit_x64.c (hunk 8 above)
Comment (jserv) [reply src/jit_x64.c:79]: Use consistent naming. No BPF should appear.

File: src/jit_x64.c [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+#include <unistd.h>
+
+#include "cache.h"
+#include "decode.h"
+#include "io.h"
+#include "jit_x64.h"
+#include "state.h"
+#include "utils.h"
+
+enum bpf_register {
```
Comment (jserv) [inline src/jit_x64.c]: Since this C source is intended for internal use, specifically for JIT code generation, the naming scheme can follow the principle of simplicity. Therefore, we can safely drop the prefix `bpf_` or `BPF_`.

File: src/jit_x64.h [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Generic x86-64 code generation functions
+ */
+
+#ifndef JIT_X64_H
```
Comment (jserv) [inline src/jit_x64.h]: Use `#pragma once` instead.

File: src/feature.h [hunk 15]
```diff
@@ -52,5 +52,10 @@
 #define RV32_FEATURE_ARC 0
 #endif
 
+/* Just-in-time compiler */
```
Comment (jserv) [inline src/feature.h]: Rephrase as "Experimental just-in-time compiler"

File: src/jit_x64.h [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "riscv_private.h"
+
+#define RAX 0
```
Comment (jserv) [inline src/jit_x64.h]: Use `enum` instead of defines for these registers.

File: src/jit_x64.h [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+    uint32_t retpoline_loc;
+    struct offset_map *offset_map;
+    int num_insn;
+    struct jump *jumps;
+    int num_jumps;
+};
+
+struct jit_state *init_state(size_t size);
+
+uint32_t translate_x86_64(riscv_t *rv, block_t *block);
```
Comment (jserv) [inline src/jit_x64.h]: Always refer to the architecture as "x64" instead of "x86-64" since we are unlikely to support 32-bit x86 for the JIT compiler. We have plans and commitments to support two targets: x64 and Arm64.

File: src/cache.c [hunk 18]
```diff
@@ -3,20 +3,21 @@
  * "LICENSE" for information on usage and redistribution of this file.
  */
 
+#include "cache.h"
```
Comment (jserv) [inline src/cache.c]: Don't change the order of local header inclusion.

File: src/jit_x64.c [hunk 19]
```diff
@@ -0,0 +1,745 @@
+// Copyright (c) 2015 Big Switch Networks, Inc
```
Comment (jserv) [inline src/jit_x64.c]: Drop the original licensing declaration since we are reworking the ubpf implementation. We can mention the following:
> This JIT implementation has undergone extensive modifications, heavily relying on the ubpf_jit_x86_64.[ch] from ubpf. The original ubpf_jit_x86_64.[ch] file served as the foundation and source of inspiration for adapting and tailoring it specifically for this JIT implementation. Therefore, credit and sincere thanks are extended to ubpf for their invaluable work.

Check src/map.[ch] for the way in which we provide credit for the previous work.

File: src/jit_x64.h [hunk 20]
```diff
@@ -0,0 +1,436 @@
+// Copyright (c) 2015 Big Switch Networks, Inc
```
Comment (jserv) [inline src/jit_x64.h]: Ditto. Rework the licensing notice and credit information.

File: src/jit_x64.h [hunk 21]
```diff
@@ -0,0 +1,436 @@
...
+{
+    emit_basic_rex(state, size == S64, dst, src);
+    if (size == S8 || size == S16) {
+        /* movzx */
+        emit1(state, 0x0f);
+        emit1(state, size == S8 ? 0xb6 : 0xb7);
+    } else if (size == S32) {
+        /* mov */
+        emit1(state, 0x8b);
+    } else if (size == S64) {
```
Comment (jserv) [inline src/jit_x64.h]: Shall we care about S64 case?

File: src/jit_x64.c (hunk 7 above)
Comment (jserv) [reply src/jit_x64.c]: Apply consistent naming scheme. That is,
```c
enum vm_reg {
    VM_REG0 = 0,
    VM_REG1,
    ...
    VM_REG10,
    N_VM_REGS,
};
```

File: src/emulate.c [hunk 22]
```diff
@@ -1030,11 +924,33 @@ void rv_step(riscv_t *rv, int32_t cycles)
             }
         }
         last_pc = rv->PC;
-
-        /* execute the block */
+#if RV32_HAS(JIT)
+        /* execute by tiered 1 JIT compiler */
```
Comment (jserv) [inline src/emulate.c]: The correct term is "tier 1 JIT compiler" and "tiered compilation."

File: src/jit_x64.c [hunk 23]
```diff
@@ -0,0 +1,745 @@
...
+#define EBPF_MODE_JSGT 0x60
+#define EBPF_MODE_JSGE 0x70
+#define EBPF_MODE_CALL 0x80
+#define EBPF_MODE_EXIT 0x90
+#define EBPF_MODE_JLT 0xa0
+#define EBPF_MODE_JLE 0xb0
+#define EBPF_MODE_JSLT 0xc0
+#define EBPF_MODE_JSLE 0xd0
+
+#define EBPF_OP_JA (EBPF_CLS_JMP | EBPF_MODE_JA)
```
Comment (jserv) [inline src/jit_x64.c]: Drop unused macros. Check if the following macros are required or not.

File: src/rv32_template.c [hunk 24]
```diff
@@ -3,43 +3,85 @@
  * specification version 20191213.
  */
 
+/* Please check files gen-jit-template.py and jit_x64.[ch] to understand custom
```
Comment (jserv) [inline src/rv32_template.c]: Improve the documentation.

**Review by jserv — CHANGES_REQUESTED** [review]
Improve the git commit messages to explicitly highlight the performance comparison between the rv32emu interpreter, QEMU, and the proposed tier-1 JIT compiler. This comparison will focus on execution time and resource utilization, with particular attention to memory consumption.

File: src/emulate.c [hunk 25]
```diff
@@ -292,9 +166,14 @@ static block_t *block_alloc(riscv_t *rv)
...
     block->n_insn = 0;
     block->predict = NULL;
+#if RV32_HAS(JIT)
+    block->hot = false;
+    block->backward = false;
+#endif
     return block;
 }
 
+#if !RV32_HAS(JIT)
```
Comment (jserv) [inline src/emulate.c:309]: Function `map_hash` is only used by interpreter.

**Comment by jserv** [comment]
This implementation has been successfully tested on x86-based macOS 10.15.7. However, note that, in compliance with Apple's security policy, additional entitlements are necessary to permit the execution of JIT-compiled code. See https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_allow-jit

Quoted from [JDK-8234930](https://bugs.openjdk.org/browse/JDK-8234930) "Use `MAP_JIT` when allocating pages for code cache on macOS":
> This patch is a strong prerequest for mac-aarch64 port. it was made slightly different to upstream patch so on platforms other than macos this patch equals to "nothing".

**Comment by qwe661234** [comment]
> Improve the git commit messages to explicitly highlight the performance comparison between the rv32emu interpreter, QEMU, and the proposed tier-1 JIT compiler. This comparison will focus on execution time and resource utilization, with particular attention to memory consumption.

Currently, we don't have effective memory examination tool to examinate the real memory usage. The memory consumption we can gain is the result without counting `mmap`.

File: src/riscv.c [hunk 26]
```diff
@@ -143,10 +155,15 @@ bool rv_enables_to_output_exit_code(riscv_t *rv)
...
 }
 
+
 void rv_delete(riscv_t *rv)
 {
     assert(rv);
+#if !RV32_HAS(JIT)
     block_map_destroy(rv);
+#else
+    cache_free(rv->block_cache, NULL);
```
Comment (jserv) [inline src/riscv.c]: The second parameter is set to `NULL`, but there is no check for callback in the implementation of `cache_free`, resulting in segmentation fault.
Send another pull request to fix prior to the merge of this pull request.

**Comment by jserv** [comment]
> Currently, we don't have effective memory examination tool to examinate the real memory usage. The memory consumption we can gain is the result without counting `mmap`.

Simply show what you have in git commit messages where we tend to track the development progress.

File: src/jit_x64.c [hunk 27]
```diff
@@ -0,0 +1,594 @@
...
+    VM_REG_2,
+    VM_REG_3,
+    VM_REG_4,
+    VM_REG_5,
+    VM_REG_6,
+    VM_REG_7,
+    VM_REG_8,
+    VM_REG_9,
+    VM_REG_10,
+    _VM_REG_MAX,
```
Comment (jserv) [inline src/jit_x64.c:47]: Replace `_VM_REG_MAX` with `N_VM_REGS` for consistent naming scheme.

File: src/jit_x64.c [hunk 28]
```diff
@@ -0,0 +1,594 @@
...
+/* This JIT implementation has undergone extensive modifications, heavily
+ * relying on the ubpf_jit_x86_64.[ch] from ubpf. The original
+ * ubpf_jit_x86_64.[ch] file served as the foundation and source of inspiration
+ * for adapting and tailoring it specifically for this JIT implementation.
+ * Therefore, credit and sincere thanks are extended to ubpf for their
+ * invaluable work.
+ * Reference:
+ *   https://github.com/iovisor/ubpf/blob/main/vm/ubpf_jit_x86_64.c
+ */
+
```
Comment (jserv) [inline src/jit_x64.c:16]: Add compilation guard. That is,
```c
#if !defined(__x86_64__)
#error "This implementation is dedicated to x86-64."
#endif
```

See https://github.com/sysprog21/kvm-host/blob/master/src/arch/x86/vm.c as an example.

File: src/jit_x64.c [hunk 29]
```diff
@@ -0,0 +1,594 @@
...
+#define X64_OP_DIV_IMM (X64_CLS_ALU | X64_SRC_IMM | 0x30)
+#define X64_OP_DIV_REG (X64_CLS_ALU | X64_SRC_REG | 0x30)
+#define X64_OP_MOD_IMM (X64_CLS_ALU | X64_SRC_IMM | 0x90)
+#define X64_OP_MOD_REG (X64_CLS_ALU | X64_SRC_REG | 0x90)
+
+#define STACK_SIZE 512
+#define MAX_INSTS 65536
+
+#if !defined(_countof)
+#define _countof(array) (sizeof(array) / sizeof(array[0]))
```
Comment (jserv) [inline src/jit_x64.c]: Use macro `ARRAYS_SIZE` defined in `src/common.h` instead.

File: src/jit_x64.c [hunk 30]
```diff
@@ -0,0 +1,594 @@
...
+#endif
+
+#if RV32_HAS(EXT_M)
+static void muldivmod(struct jit_state *state,
```
Comment (jserv) [inline src/jit_x64.c:71]: Is this forward declaration required?

File: src/jit_x64.c [hunk 31]
```diff
@@ -0,0 +1,594 @@
...
+ * There are two common x86-64 calling conventions, as discussed at
+ * https://en.wikipedia.org/wiki/X64_calling_conventions#x86-64_calling_conventions
+ *
+ * Please Note: R12 is special and we are *not* using it. As a result, it is
+ * omitted from the list of non-volatile registers for both platforms (even
+ * though it is, in fact, non-volatile).
+ */
+
+#if defined(_WIN32)
+static int nonvolatile_reg[] = {RBP, RBX, RDI, RSI, R13, R14, R15};
```
Comment (jserv) [inline src/jit_x64.c:215]: Add `const` qualifier when possible.

File: src/emulate.c [hunk 32]
```diff
@@ -978,6 +987,10 @@ static block_t *block_find_or_translate(riscv_t *rv)
     return next;
 }
 
+#if RV32_HAS(JIT)
+typedef void (*exec_tired1_block_func_t)(riscv_t *rv, uintptr_t);
```
Comment (jserv) [inline src/emulate.c]: Replace `exec_tired1_block_func_t ` with `exec_block_func_t ` because we don't have to distinguish from tier-1 and tier-2 compilers at the moment.

Comment (jserv) [reply src/emulate.c]: In addition, "tired1" is considered as a typo.

**Comment by jserv** [comment]
When creating tables or performance metrics, you can simply name this Tier-1 JIT compiler as "rv32emu-T1C" or "T1C."

File: src/emulate.c [hunk 33]
```diff
@@ -370,8 +382,14 @@ static bool is_branch_taken = false;
 /* record the program counter of the previous block */
 static uint32_t last_pc = 0;
 
+#if RV32_HAS(JIT)
+/* record whether the block is replaced by cache. If so, clear the EBB
```
Comment (jserv) [inline src/emulate.c]: Refine the use of "EBB" since we move to block chaining.

**Review by jserv — CHANGES_REQUESTED** [review]
Update CI pipeline to validate the JIT compiler.

**Comment by jserv** [comment]
Consistently utilize the phrase "tier-1 JIT compiler" or "tier-1 compiler," and you can opt for the shortened form "T1C" for brevity.

File: src/emulate.c [hunk 34]
```diff
@@ -1030,11 +1051,33 @@ void rv_step(riscv_t *rv, int32_t cycles)
...
+        struct jit_state *state = rv->jit_state;
+        if (block->hot) {
+            ((exec_block_func_t) state->buf)(
+                rv, (uintptr_t) (state->buf + block->offset));
+            prev = NULL;
+            continue;
+        } /* check if using frequency of block exceed threshold */
+        else if ((block->backward &&
+                  cache_freq(rv->block_cache, block->pc_start) >= 1024) ||
+                 (cache_hot(rv->block_cache, block->pc_start))) {
```
Comment (jserv) [inline src/emulate.c]: Remove the parentheses out of `cache_hot`.

File: src/jit_x64.c [hunk 35]
```diff
@@ -0,0 +1,593 @@
...
+    VM_REG_2,
+    VM_REG_3,
+    VM_REG_4,
+    VM_REG_5,
+    VM_REG_6,
+    VM_REG_7,
+    VM_REG_8,
+    VM_REG_9,
+    VM_REG_10,
+    _VM_REG_MAX,
```
Comment (jserv) [inline src/jit_x64.c]: Replace `_VM_REG_MAX` with `N_VM_REGS` for consistent naming scheme.

File: src/jit_x64.c [hunk 36]
```diff
@@ -0,0 +1,593 @@
...
+#define X64_SRC_REG 0x08
+#define X64_OP_MUL_IMM (X64_CLS_ALU | X64_SRC_IMM | 0x20)
+#define X64_OP_MUL_REG (X64_CLS_ALU | X64_SRC_REG | 0x20)
+#define X64_OP_DIV_IMM (X64_CLS_ALU | X64_SRC_IMM | 0x30)
+#define X64_OP_DIV_REG (X64_CLS_ALU | X64_SRC_REG | 0x30)
+#define X64_OP_MOD_IMM (X64_CLS_ALU | X64_SRC_IMM | 0x90)
+#define X64_OP_MOD_REG (X64_CLS_ALU | X64_SRC_REG | 0x90)
+
+#define STACK_SIZE 512
+#define MAX_INSTS 1024
```
Comment (jserv) [inline src/jit_x64.c]: Replace `MAX_INSTS` with `MAX_INSNS` for consistency.

File: src/jit_x64.c [hunk 37]
```diff
@@ -0,0 +1,593 @@
...
+#define _(inst, can_branch, insn_len, reg_mask) [rv_insn_##inst] = do_##inst,
+    RV_INSN_LIST
+#undef _
+    /* Macro operation fusion instructions */
+#define _(inst) [rv_insn_##inst] = do_##inst,
+    FUSE_INSN_LIST
+#undef _
+};
+/* clang-format on */
+typedef void (*codegen_tired1_block_func_t)(struct jit_state *,
```
Comment (jserv) [inline src/jit_x64.c]: Replace `codegen_tired1_block_func_t` with `codegen_block_func_t`.

File: src/rv32_template.c [hunk 38]
```diff
@@ -3,43 +3,86 @@
  * specification version 20191213.
  */
 
+/* Please review the files gen-jit-template.py and jit_x64.[ch] to comprehend
```
Comment (jserv) [inline src/rv32_template.c]: Rather than specifying "Please review the files gen-jit-template.py and jit_x64.[ch]," it is more beneficial to provide an overview of the domain-specific language (DSL) specification for JIT code generation. Documentation serves the purpose of simplifying maintenance tasks, rather than merely serving as an overly simplified reference.

**Comment by jserv** [comment]
The proposed JIT compiler does not encompass effective code generation for RV32F. You should clarify which instruction sets and extensions are supported.

While launching the Quake via `make quake`, I am aware of the following:
```
Playing registered version.
Console initialized.
Exe: 20:12:56 Sep  1 2023
 8.0 megabyte heap
792k surface cache
Assertion failed: (NULL), function do_fmvwx, file src/rv32_jit_template.c, line 556.
```

Another reproducer:
```
 $ build/rv32emu build/maj2random.elf
maj_random (NROUNDS=2)                 count     mean(x)     mean(y) variance(x) variance(y)  std-dev(x)  std-dev(y)
(             0, (0 - 1K)/8K )    999999,057     0.49915     0.50594     0.08122     0.08092     0.28499     0.28447
Assertion failed: (NULL), function do_fmvwx, file src/rv32_jit_template.c, line 556.
```

**Comment by jserv** [comment]
Tested on Ubuntu 20.04, Python 3.8 generates the following error messages:
```
  File "tools/gen-jit-template.py", line 153
    match items[0]:
          ^
SyntaxError: invalid syntax
make: *** [Makefile:130: src/rv32_jit_template.c] Error 1
$ python3 --version
Python 3.8.10
```

I installed Python 3.9.5 for comparison, and the above message still appear.
According to [Status of Python versions](https://devguide.python.org/versions/), Python 3.8 and 3.9 are still supported by Python Software Foundation.

**Comment by qwe661234** [comment]
> Tested on Ubuntu 20.04, Python 3.8 generates the following error messages:
> 
> ```
>   File "tools/gen-jit-template.py", line 153
>     match items[0]:
>           ^
> SyntaxError: invalid syntax
> make: *** [Makefile:130: src/rv32_jit_template.c] Error 1
> $ python3 --version
> Python 3.8.10
> ```
> 
> I installed Python 3.9.5 for comparison, and the above message still appear. According to [Status of Python versions](https://devguide.python.org/versions/), Python 3.8 and 3.9 are still supported by Python Software Foundation.

As alternative, I rewrite python switch into if-else.

**Comment by jserv** [comment]
The JIT-compiled code can initiate Doom, but it crashes when attempting to start the game.

Tested on Ubuntu Linux 20.04.

File: src/rv32_template.c [hunk 39]
```diff
@@ -3,43 +3,123 @@
  * specification version 20191213.
  */
 
+/* Currently, the tier-1 JIT compiler only supports for RV32-IMC extension,
+ * RV32-A and RV32-F extension are unsupported.
+ */
+
+/* size: size of data
```
Comment (jserv) [inline src/rv32_template.c]: Add general descriptions prior to the following. Make them informative.

**Comment by jserv** [comment]
While T1C is capable of running Quake, it does not behave as expected when compared to the interpreter. I encountered an unexpected crash with Quake. To reproduce this, first run `make quake` and watch the demo scene without any keystrokes. You will then encounter a segmentation fault.
```
You got the shells
You got the Grenade Launcher
You receive 25 health
You get 2 rockets  
You got the rockets
You got the nailgun
You got the nails
You got the nails
You receive 15 health
You receive 15 health
You got the rockets
/bin/sh: line 1: 71739 Segmentation fault: 11  ../build/rv32emu quake.elf
```

Th log generated by the interpreter:
```
You got the shells
You got the Grenade Launcher
You receive 25 health
You get 2 rockets  
You got the rockets
You got the nailgun
You got the nails
You got the nails
You receive 15 health
You receive 15 health
You got the rockets
You get 2 rockets  
You got the nails
You got the nails
You receive 25 health
You got the gold key
Playing demo from demo2.dem.
You got the shells
You got the shells
You got the nails
You got armor
You got the nails
You receive 25 health
You got the Super Nailgun
You got the nails
You get 2 rockets  
```

**Comment by qwe661234** [comment]
> While T1C is capable of running Quake, it does not behave as expected when compared to the interpreter. I encountered an unexpected crash with Quake. To reproduce this, first run `make quake` and watch the demo scene without any keystrokes. You will then encounter a segmentation fault.
> 
> ```
> You got the shells
> You got the Grenade Launcher
> You receive 25 health
> You get 2 rockets  
> You got the rockets
> You got the nailgun
> You got the nails
> You got the nails
> You receive 15 health
> You receive 15 health
> You got the rockets
> /bin/sh: line 1: 71739 Segmentation fault: 11  ../build/rv32emu quake.elf
> ```
> 
> Th log generated by the interpreter:
> 
> ```
> You got the shells
> You got the Grenade Launcher
> You receive 25 health
> You get 2 rockets  
> You got the rockets
> You got the nailgun
> You got the nails
> You got the nails
> You receive 15 health
> You receive 15 health
> You got the rockets
> You get 2 rockets  
> You got the nails
> You got the nails
> You receive 25 health
> You got the gold key
> Playing demo from demo2.dem.
> You got the shells
> You got the shells
> You got the nails
> You got armor
> You got the nails
> You receive 25 health
> You got the Super Nailgun
> You got the nails
> You get 2 rockets  
> ```

Because quake has RV32F insturction, but jit does not support it now.

**Comment by jserv** [comment]
> Because quake has RV32F instructions, but T1C does not support it now.

We do need an effective fallback for RV32F and RV32A when the JIT compiler is unable to generate the corresponding machine code. A trampoline would be required for switching between translated blocks and non-JIT parts.
See https://github.com/jserv/rv32jit/blob/master/src/codegen/jitabi.cpp

File: src/decode.h [hunk 40]
```diff
@@ -32,150 +32,150 @@ enum op_field {
...
-        _(cbnez, 1, 2, ENC(rs1))                    \
-        _(cslli, 0, 2, ENC(rd))                     \
-        _(clwsp, 0, 2, ENC(rd))                     \
-        _(cjr, 1, 2, ENC(rs1, rs2, rd))             \
-        _(cmv, 0, 2, ENC(rs1, rs2, rd))             \
-        _(cebreak, 1, 2, ENC(rs1, rs2, rd))         \
-        _(cjalr, 1, 2, ENC(rs1, rs2, rd))           \
-        _(cadd, 0, 2, ENC(rs1, rs2, rd))            \
-        _(cswsp, 0, 2, ENC(rs2))                    \
+#define RV_INSN_LIST                                   \
```
Comment (jserv) [inline src/decode.h:35]: Refine the above comment for new field.

**Comment by jserv** [comment]
Finally, the JIT compiler is incorporated into this emulator repository.
